c/src/point_one/polaris/polaris.c

```diff
@@ -1311,40 +1360,57 @@ static int GetHTTPResponse(PolarisContext_t* context) {
     size_t content_length =
         total_bytes - (content_start - (char*)context->recv_buffer);
     memmove(context->recv_buffer, content_start, content_length);
-    P1_DebugPrint("Response content:\n%s\n", context->recv_buffer);
+    P1_PrintDebug("Response content:\n%s", context->recv_buffer);
   } else {
     // No content in response.
-    P1_DebugPrint("No content in response.\n");
+    P1_PrintDebug("No content in response.");
     context->recv_buffer[0] = '\0';
   }
 
   return status_code;
 }
 
 /******************************************************************************/
-#if !defined(P1_NO_PRINT)
+#if !P1_NO_PRINT
 void P1_PrintData(const uint8_t* buffer, size_t length) {
   if (__log_level < POLARIS_LOG_LEVEL_TRACE) {
     return;
   }
 
+  char str[64] = {'\0'};
+  int str_length = 0;
   for (size_t i = 0; i < length; ++i) {
     if (i % 16 != 0) {
-      P1_fprintf(stderr, " ");
+      str[str_length++] = ' ';
     }
 
-    P1_fprintf(stderr, "%02x", buffer[i]);
+    str_length +=
+        snprintf(str + str_length, sizeof(str - str_length), "%02x", buffer[i]);
 
     if (i % 16 == 15) {
-      P1_fprintf(stderr, "\n");
+      str[str_length++] = '\0';
+      if (__print_callback) {
+        __print_callback("polaris.c", __LINE__, POLARIS_LOG_LEVEL_TRACE, str);
+      } else {
+        P1_fprintf(stderr, "%s", str);
+      }
+      str_length = 0;
+    }
+  }
+
+  if (str_length != 0) {
+    str[str_length++] = '\0';
+    if (__print_callback) {
+      __print_callback("polaris.c", __LINE__, POLARIS_LOG_LEVEL_TRACE, str);
```

Comment:
if using line, we should use __FILE__

Reply:
* the above should be the c style thing with two underscore in front and behind, github is formatting it instead lol

Reply:
In theory, yes. We did this just to simplify/shorten the print since `__FILE__` always includes the full path to the file (`src/...`) and this is the only source file in the repo that prints anything. Unfortunately there's no macro to get just the filename, and you can't do string manipulation in the pre-processor.

I can switch it to `__FILE__` if you think that's better?